### PR TITLE
Move unit tests context to platform heap instead of stack/static

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
Test contexts only need to live for the duration of the test:
  - on stack means extra stack usage (if objects grow too much this is problematic - specifically #19819 overflows zephyr stack)
  - static means .bss usage, which is shared on platforms such as zephyr

#### Change overview
- move test support helpers in separate library
- make all tests link in the test support library
- add "run with context" and "run without context" functions for executing tests
- make use of these functions in unit tests that are using contexts

The following had complex logic (or using globals) and were NOT updated:

- `TestSystemPacketBuffer.cpp`
- `TestAdvertiser.cpp`
- `TestDataModelSerialization.cpp`

#### Testing
Unit test still pass

### CODE REVIEW

**Review by tcarmelveilleux — APPROVED** [review]

**Comment by andy31415** [comment]
fast track because unit test updates (i.e. 100% unit tested) and looking to unblock #19819

**Review by bzbarsky-apple — DISMISSED** [review]

File: src/protocols/secure_channel/CASEServer.h
Comment (andy31415) [inline src/protocols/secure_channel/CASEServer.h:47]: @bzbarsky-apple @mrjerryjohns 

Made this change to fix ordering issue:
  - if CASEServer has a longer lifetime than the exchange manager associated with it (like the tests), then both Shutdown is called "normally" and in the destructor
  - by the time the destructor is called, the exchange manager may already be invalid

File: src/crypto/CHIPCryptoPALOpenSSL.cpp
Comment (andy31415) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:1819]: @emargolis @tcarmelveilleux - this seemed a leak on object allocation from text. Changed to make asan tests pass for me, but unsure why CI never saw these,

File: src/lib/core/tests/TestCHIPTLV.cpp
Comment (andy31415) [inline src/lib/core/tests/TestCHIPTLV.cpp:246]: @mrjerryjohns  - this was an API misunderstanding it seems, both for DupString and DupBytes.

I do not see usages of DupString/Bytes beyond tests. Should we instead remove these methods?

File: src/protocols/secure_channel/CASEServer.h
Comment (bzbarsky-apple) [reply src/protocols/secure_channel/CASEServer.h:47]: CASEServer must not outlive the exchange manager.  Why is that happening in tests?  Can we fix them to not do that?

The shutdown change is ok, I think, but the lifetime ordering still needs to be right.

File: src/crypto/CHIPCryptoPALOpenSSL.cpp
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:1819]: @andy31415 Why is #19819 blocked on this?  I thought we bumped the stack size to avoid this problem there?

I'm not comfortable with fast-tracking these sorts of changes.  Furthermore, I think the non-test changes here should be broken out into a separate PR so they don't get lost in the huge test diff.  I'd also really like to understand what made these non-test issues come to light in this PR....

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/crypto/CHIPCryptoPALOpenSSL.cpp
Comment (andy31415) [reply src/crypto/CHIPCryptoPALOpenSSL.cpp:1819]: It spiraled:
   - 19189 was running out of stack because test contexts got larger, so Tennessee asked me to create a PR to move all contexts to heap
   - this exposed a ASAN failure (the CASESession accessing a freed ExchangeManager because TestCASESession is global and the contexts were moved to heap)
   - this got me running is_asan locally and that seems to leak detect too for me (maybe because not clang, not sure, but it looks like legitimate leaks)
   
 So now I have:
   - move contexts to heap
   - changes to fix memory leaks in tests and this is the only non-test one

Comment (andy31415) [reply src/crypto/CHIPCryptoPALOpenSSL.cpp:1819]: OpenSSL seems to leak 3 objects every time we call ExtractVIDPIDFromX509Cert. Seems bad.

Comment (andy31415) [reply src/crypto/CHIPCryptoPALOpenSSL.cpp:1819]: I removed the fast track - originally added it because it was "test only" changes. But tests found bugs too

Comment (andy31415) [reply src/crypto/CHIPCryptoPALOpenSSL.cpp:1819]: Regarding why these came out, I believe my asan settings were somehow stricter - CI found the exchange manager crash but not leaks, but locally I found leaks and fixed those too.

Comment (bzbarsky-apple) [reply src/crypto/CHIPCryptoPALOpenSSL.cpp:1819]: > this exposed a ASAN failure (the CASESession accessing a freed ExchangeManager because TestCASESession is global and the contexts were moved to heap)

I'm not following this part.  TestCASESession is a function...  Is this about `CASEServerForTest`?

If so, then we should hang it as a member off the context (which should then inherit from `Test::LoopbackMessagingContext` but have this extra member), so our lifetimes can be sane.  Or make it a pointer and allocate it in `CASETestSecurePairingSetup` and destroy in `CASE_TestSecurePairing_Teardown`.  Because having tests do differrent lifetime ordering than our real code is not very useful...

> Regarding why these came out, I believe my asan settings were somehow stricter - CI found the exchange manager crash but not leaks

I wish we could figure out why and fix CI to catch those....  Are they possibly showing up in CI but not being treated as fatal or something?  Would be good to figure out what's going on.

In any case, I think we should have a PR to fix the bugs, and a separate PR for the test mass-change.

Comment (andy31415) [reply src/crypto/CHIPCryptoPALOpenSSL.cpp:1819]: `CASEServerForTest` is a `CASEServer` which will access mExchangeManager on ::Shutdown() which is called in the destructor.

So moving a lifetime of exchange manager to be less then CASEServer will make an illegal memory access. So existing code relies on static destructor order, which is also not ok but happens to work.

File: src/protocols/secure_channel/CASEServer.h
Comment (andy31415) [reply src/protocols/secure_channel/CASEServer.h:47]: I can fix the tests, but generally the logic was likely not ok: it relied on static object lifetime before that happened to work.

I am trying to not have PRs spiral too much into yak shaving.

File: src/crypto/CHIPCryptoPALOpenSSL.cpp
Comment (andy31415) [reply src/crypto/CHIPCryptoPALOpenSSL.cpp:1819]: Split out bug fixing into https://github.com/project-chip/connectedhomeip/pull/19953

Still need to keep this PR since it requires those bug fixes to pass.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

File: src/crypto/CHIPCryptoPALOpenSSL.cpp
Comment (bzbarsky-apple) [reply src/crypto/CHIPCryptoPALOpenSSL.cpp]: @andy31415 My point was that CASEServerForTest should have a _smaller_ lifetime than the exchange manager.  And I proposed a specific way to accomplish that above....

File: src/protocols/secure_channel/CASEServer.h
Comment (bzbarsky-apple) [reply src/protocols/secure_channel/CASEServer.h]: Fwiw, the logic was generally sane-ish fine: CASEServer is inited with ExchangeManager, so you have to tear down CASEServer before tearing down EXchangeManager, and the tests were just not doing that....

Comment (bzbarsky-apple) [reply src/protocols/secure_channel/CASEServer.h]: @andy31415 Yes, I have no problem with this change.  I just think we should _also_ fix the test to do what the SDK itself does.

**Pull request merged**
